deps/3rdparty/libdispatch/src/queue.c

```diff
@@ -1978,3 +2002,68 @@ dispatch_after_f(dispatch_time_t when, dispatch_queue_t queue, void *ctxt, void
 	dispatch_resume(as_do(ds));
 }
 
+static void _dispatch_queue_init_specific_list(dispatch_queue_t dq) {
+	dispatch_queue_specific_list_t dqsq;
+	dqsq = calloc(1, sizeof(struct dispatch_queue_specific_list_s));
+	TAILQ_INIT(&dqsq->contextList);
+	dq->dq_specific_q = dqsq;
+}
+
+static void _dispatch_queue_insert_specific(dispatch_queue_t dq, dispatch_queue_specific_t dqs) {
+	dispatch_queue_specific_list_t dqsl = dq->dq_specific_q;
+	dispatch_queue_specific_t var;
+
+	TAILQ_FOREACH(var, &dqsl->contextList, specific) {
+		if (var->key == dqs->key) {
+			if(var->destructor) {
+				var->destructor(var->context);
+			}
+
+			// If a context exists for this key, simply update the context on that specific
+			if(dqs->context) {
+				var->context = dqs->context;
+				var->destructor = dqs->destructor;
+			} else {
+				TAILQ_REMOVE(&dqsl->contextList, var, specific);
+				free(var);
+			}
+			free(dqs);
+			return;
+		}
+	}
+
+	TAILQ_INSERT_TAIL(&dqsl->contextList, dqs, specific);
```

Comment:
Concern: On line 2027 you remove an *existing* specific if you set its value to null.

Here, you will insert a *new* specific _even if_ its value is null.

Reply:
Good catch. There shouldn't be a specific with a null context.